### DISCUSSION TITLE
OpenCL compilation

### DISCUSSION THREAD

**Discussion by yyssys** [sarchlab/mgpusim#58] [Q&A] [answered]
Hello author, I have encountered an issue where in some benchmarks, the original HSACO and disasm files were deleted, and the kernel function was compiled using the clang-ocl command to generate two files (as shown in the figure below, which is different from the original). However, some errors may occur during the compilation phase and others during the execution phase.What is the reason for this? How can I solve this problem? Looking forward to your reply.
![Image_1713415150704](https://github.com/sarchlab/mgpusim/assets/161021701/9b69e97b-f135-4c38-8c59-30265fb02434)
![Image_1713415146821](https://github.com/sarchlab/mgpusim/assets/161021701/01b39c31-5c09-4b31-b83c-9586f82a1743)

**Comment by syifan** [comment]
The first image shows that this is because we do not have the instruction implemented. Do you want to try to implement it? 
The second error seems like an OpenCL error. You should modify the OpenCL code to fix it. 

Which benchmark misses the assembly file and hsaco files? I think we should have all of them.

**Reply by yyssys** [reply]
Thank you for your reply. No benchmark missing assembly files and hsaco files. My puzzle is why the original HSACO file executes without error, but I recompile the kernel function without modifying any code and then execute it with an error message. I would like to know how you compiled and generated the original HSACO file. Looking forward to your reply.

**Reply by syifan** [reply, marked as answer]
For the Opcode 17 error, it is probably just because of a version difference in the compiler. We are using the rocm-dev docker image with version 3.8 to compile many kernels, but many other kernels may be compiled by even earlier compiler versions.

**Reply by yyssys** [reply]
Thank you, I will try an older compiler version to see if it can solve this problem.

**Reply by yyssys** [reply]
Hello author, I'm sorry to bother you again. I tried many compiler versions today but failed in the end. Out of interest, I wanted to do some small benchmarks on MGPUSim for learning, but the kernel function compilation problem has puzzled me for a long time. Therefore, I would like to ask if you could share the rocm-dev docker image you use, or provide some guidance so that I can get it. If you have time to share this image or provide some relevant guidance, I would be grateful.

**Reply by syifan** [reply]
This is the docker image that we have been using https://hub.docker.com/layers/rocm/dev-ubuntu-20.04/3.8/images/sha256-788b88873e18d83b3abe80a324ec106b598c7c1a949ff7fec35e295b6e6e9748?context=explore. However, it has been quite a while since the last time we compile an OpenCL kernel. If you have difficulty, perhaps I can have a meeting with you to see what is the problem?

**Reply by yyssys** [reply]
Thank you for your reply, but it is too late in Beijing now. I will try this docker image for the first time tomorrow. If I still have any problems, please feel free to advise me.

**Comment by Yzgll** [comment]
@syifan Unfortunately, I tried all versions of ROCm and none of them worked correctly after being recompiled. Then I tried other benchmarks such as fir and recompiled and ran it but it worked correctly, it seems that Opcode 17 error is only for AES. I really don't know how to solve it.
